packages/strapi/lib/Strapi.js

```diff
@@ -201,45 +201,57 @@ class Strapi extends EventEmitter {
   }
 
   async bootstrap() {
-    if (!this.config.functions.bootstrap) {
-      return Promise.resolve();
-    }
+    const bootstrapFunctions = [];
```

Comment:
Two things that I don't like in this method.

- Using forEach method to populate an array.
- The main `bootstrap` function is called at the same time as the plugins `bootstrap function`.

I think the main bootstrap function should be called at the very end of the process. Here my proposition using your code:

```js
const execBootstrap = (fn) => !fn ? Promise.resolve() : new Promise((resolve, reject) => {
  const timeoutMs = this.config.bootstrapTimeout || 3500;
  const timer = setTimeout(() => {
    this.log.warn(`Bootstrap is taking unusually long to execute its callback ${timeoutMs} miliseconds).`);
    this.log.warn('Perhaps you forgot to call it?');
  }, timeoutMs);

  let ranBootstrapFn = false;

  try {
    fn(err => {
      if (ranBootstrapFn) {
        this.log.error('You called the callback in `strapi.config.boostrap` more than once!');

        return reject();
      }

      ranBootstrapFn = true;
      clearTimeout(timer);

      return resolve(err);
    });
  } catch (e) {
    if (ranBootstrapFn) {
      this.log.error('The bootstrap function threw an error after its callback was called.');

      return reject(e);
    }

    ranBootstrapFn = true;
    clearTimeout(timer);

    return resolve(e);
  }
});

return Promise.all(
  Object.values(this.plugins)
  .map(x => execBootstrap(get(x, 'config.functions.bootstrap')))
).then(() => execBootstrap(this.config.functions.bootstrap));
```